apiserver/common/charms/common.go

```diff
@@ -10,130 +10,199 @@ import (
+func NewApplicationCharmInfoAPI(modelTag names.ModelTag, service ApplicationService, authorizer facade.Authorizer) (*ApplicationCharmInfoAPI, error) {
 	return &ApplicationCharmInfoAPI{
+		modelTag:   modelTag,
 		authorizer: authorizer,
-		state:      st,
+		service:    service,
 	}, nil
 }
 
 // ApplicationCharmInfo fetches charm information for an application.
 func (a *ApplicationCharmInfoAPI) ApplicationCharmInfo(ctx context.Context, args params.Entity) (params.Charm, error) {
-	if err := checkCanRead(ctx, a.authorizer, a.state); err != nil {
+	if err := checkCanRead(ctx, a.authorizer, a.modelTag); err != nil {
 		return params.Charm{}, errors.Trace(err)
 	}
 
 	appTag, err := names.ParseApplicationTag(args.Tag)
 	if err != nil {
 		return params.Charm{}, errors.Trace(err)
 	}
-	app, err := a.state.Application(appTag.Id())
+
+	// Application name is used to fetch the charm information.
+	appName := appTag.Id()
+
+	ch, origin, platform, err := a.service.GetCharmByApplicationName(ctx, appName)
+	if errors.Is(err, applicationerrors.ApplicationNotFound) {
+		return params.Charm{}, errors.NotFoundf("application %q not found", appName)
+	} else if errors.Is(err, applicationerrors.CharmNotFound) {
+		return params.Charm{}, errors.NotFoundf("charm for application %q not found", appName)
+	} else if err != nil {
+		return params.Charm{}, errors.Trace(err)
+	}
+
+	return convertCharm(appName, ch, origin, platform)
+}
+
+func convertSource(source applicationcharm.CharmSource) (string, error) {
+	switch source {
+	case applicationcharm.CharmHubSource:
+		return "ch", nil
+	case applicationcharm.LocalSource:
+		return "local", nil
+	default:
+		return "", errors.Errorf("unsupported source %q", source)
+	}
+}
+
+func convertApplication(arch application.Architecture) (string, error) {
+	switch arch {
+	case applicationcharm.AMD64:
+		return "amd64", nil
+	case applicationcharm.ARM64:
+		return "arm64", nil
+	case applicationcharm.PPC64EL:
+		return "ppc64el", nil
+	case applicationcharm.S390X:
+		return "s390x", nil
+	case applicationcharm.RISV64:
+		return "riscv64", nil
+	default:
+		return "", errors.Errorf("unsupported architecture %q", arch)
+	}
+}
+
+func convertCharm(
+	name string, ch charm.Charm,
+	origin applicationcharm.CharmOrigin,
+	platform applicationcharm.Platform,
+) (params.Charm, error) {
+	schema, err := convertSource(origin.Source)
 	if err != nil {
 		return params.Charm{}, errors.Trace(err)
 	}
-	ch, _, err := app.Charm()
+
+	architecture, err := convertApplication(platform.Architecture)
 	if err != nil {
 		return params.Charm{}, errors.Trace(err)
 	}
-	return convertCharm(ch), nil
-}
 
-func convertCharm(ch Charm) params.Charm {
-	charm := params.Charm{
-		Revision: ch.Revision(),
-		URL:      ch.URL(),
+	url := charm.URL{
+		Schema:       schema,
+		Name:         name,
+		Revision:     origin.Revision,
+		Architecture: architecture,
+	}
```

Comment:
This was a right pain to ensure it was correct. We're now having to write a lot more data just to preserve the natural key.